matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/verification/SasTransportToDevice.kt

```diff
@@ -0,0 +1,140 @@
+                            }
+                        }
+
+                        override fun onFailure(failure: Throwable) {
+                            Timber.e("## SAS verification [$tx.transactionId] failed to send toDevice in state : $tx.state")
+
+                            tx.cancel(onErrorReason)
+                        }
+                    }
+                }
+                .executeBy(taskExecutor)
+    }
+
+    override fun done(transactionId: String) {
+        // To device do not do anything here
+    }
+
+    override fun cancelTransaction(transactionId: String, userId: String, userDevice: String, code: CancelCode) {
+        Timber.d("## SAS canceling transaction $transactionId for reason $code")
+        val cancelMessage = KeyVerificationCancel.create(transactionId, code)
+        val contentMap = MXUsersDevicesMap<Any>()
+        contentMap.setObject(userId, userDevice, cancelMessage)
+        sendToDeviceTask
+                .configureWith(SendToDeviceTask.Params(EventType.KEY_VERIFICATION_CANCEL, contentMap, transactionId)) {
+                    this.callback = object : MatrixCallback<Unit> {
+                        override fun onSuccess(data: Unit) {
+                            Timber.v("## SAS verification [$transactionId] canceled for reason ${code.value}")
+                        }
+
+                        override fun onFailure(failure: Throwable) {
+                            Timber.e(failure, "## SAS verification [$transactionId] failed to cancel.")
+                        }
+                    }
+                }
+                .executeBy(taskExecutor)
+    }
+
+    override fun createAccept(tid: String,
+                              keyAgreementProtocol: String,
+                              hash: String,
+                              commitment: String,
+                              messageAuthenticationCode: String,
+                              shortAuthenticationStrings: List<String>)
+            : VerificationInfoAccept = KeyVerificationAccept.create(
+            tid,
+            keyAgreementProtocol,
+            hash,
+            commitment,
+            messageAuthenticationCode,
+            shortAuthenticationStrings)
+
+    override fun createKey(tid: String, pubKey: String): VerificationInfoKey = KeyVerificationKey.create(tid, pubKey)
+
+    override fun createMac(tid: String, mac: Map<String, String>, keys: String) = KeyVerificationMac.create(tid, mac, keys)
+
+    override fun createStart(fromDevice: String,
+                             method: String,
+                             transactionID: String,
+                             keyAgreementProtocols: List<String>,
+                             hashes: List<String>,
+                             messageAuthenticationCodes: List<String>,
+                             shortAuthenticationStrings: List<String>): VerificationInfoStart {
+        return KeyVerificationStart().apply {
```

Comment:
KeyVerificationStart (and maybe other classes for Key verification) should be a data class with val defined in the constructor?

Reply:
done